### PR TITLE
TFactor: more test cases + improvements

### PR DESCRIPTION
**Pull request by albestro**
This PR introduces for the TFactor algorithm:
- usage of the `executor_mpi`
- ~reduce in-place~
- fixes for edge cases and pre-conditions

Moreover, its test has been extended with more test-cases (they might be too much, look forward for your feedback).

This PR depends on
- [x] #330 
- [x] #374 

EDIT:
1. BREAKING CHANGE: this PR changes the API of the function. Indeed, it now accepts a `Pipeline<Communicator>` in place of the old `Pipeline<CommunicatorGrid>`.
2. Computations are now performed on the HIGH PRIORITY DLAF executor
3. The communication is now asynchronous

EDIT2:
1. Instead of having multiple TRMV tasks, now they are merged all together (the for loop has been moved inside the task)

### CODE REVIEW

**Comment by albestro** [comment]
There are some minor conflicts that are going to be easily solved once all dependencies of this branch will be merged (and then I can do a final rebase onto the new master)

File: test/unit/factorization/mc/test_compute_t_factor.cpp [hunk 1]
```diff
@@ -327,14 +368,19 @@ TYPED_TEST(ComputeTFactorTestMC, CorrectnessDistributed) {
...
 
-      Matrix<TypeParam, Device::CPU> t_output({k, k}, block_size);
+      const auto max_k = std::min(mb, nb);
+      Matrix<TypeParam, Device::CPU> t_output({max_k, max_k}, block_size);
       const LocalTileIndex t_idx(0, 0);
 
       dlaf::factorization::internal::computeTFactor<Backend::MC>(k, v_input, v_start, taus_input,
                                                                  t_output(t_idx), serial_comm);
 
+      // No reflectors, so nothing to check
```
[the remark this answers is not in the document]
Comment (rasolca) [reply test/unit/factorization/mc/test_compute_t_factor.cpp:380]: I would check it.

File: test/unit/factorization/mc/test_compute_t_factor.cpp [hunk 2]
```diff
@@ -174,10 +179,35 @@ MatrixLocal<T> computeHFromTFactor(const SizeType k, const Tile<const T, Device:
   return h_result;
 }
 
-std::vector<std::tuple<SizeType, SizeType, SizeType, SizeType, SizeType>> configs{
+std::vector<std::tuple<SizeType, SizeType, SizeType, SizeType, SizeType, GlobalTileIndex>> configs{
     // m, n, mb, nb, k
```
Comment (rasolca) [inline test/unit/factorization/mc/test_compute_t_factor.cpp:183]: The legend hasn't been updated. What is the new value?

File: test/unit/factorization/mc/test_compute_t_factor.cpp [hunk 3]
```diff
@@ -174,10 +179,35 @@ MatrixLocal<T> computeHFromTFactor(const SizeType k, const Tile<const T, Device:
...
+    // non-square tiles
+    {39, 26, 10, 4, 4, {0, 5}},  // all reflectors (complete tile)
+    {39, 26, 10, 4, 2, {0, 6}},  // all reflectors (incomplete tile)
+    {39, 26, 10, 4, 3, {0, 5}},  // k reflectors (complete tile)
+    {39, 26, 10, 4, 1, {0, 6}},  // k reflectors (incomplete tile)
+
+    {26, 39, 10, 6, 6, {0, 5}},  // all reflectors (complete tile)
+    {26, 39, 10, 6, 2, {0, 6}},  // all reflectors (incomplete tile)
+    {26, 39, 10, 6, 4, {0, 5}},  // k reflectors (complete tile)
+    {26, 39, 10, 6, 1, {0, 6}},  // k reflectors (incomplete tile)
```
Comment (rasolca) [inline test/unit/factorization/mc/test_compute_t_factor.cpp:203]: I would test cases in which the row index is >0 as well.

File: test/unit/factorization/mc/test_compute_t_factor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (albestro) [reply test/unit/factorization/mc/test_compute_t_factor.cpp:380]: `computeTFactor` can alter just the tile used for storing the TFactor, since the other parameters are readonly, and in this case it means that the original tile is exactly as it was before the call (no reflectors, no tile for tfactor).

Is it what you'd like to check or were you thinking about something more?

File: test/unit/factorization/mc/test_compute_t_factor.cpp [hunk 4]
```diff
@@ -18,6 +18,7 @@
 
 #include "dlaf/common/range2d.h"
 #include "dlaf/communication/communicator_grid.h"
+#include "dlaf/lapack/tile.h"  // workaround for importing lapack.hh
```
Comment (msimberg) [inline test/unit/factorization/mc/test_compute_t_factor.cpp:21]: Why is this a workaround, or rather why is a workaround needed?

**Review by msimberg — COMMENTED** [review]
I've no particularly useful comments on this, so just commenting to say that I've looked through it and didn't notice anything terrible.

File: test/unit/factorization/mc/test_compute_t_factor.cpp (hunk 4 above)
Comment (rasolca) [reply test/unit/factorization/mc/test_compute_t_factor.cpp:21]: lapackpp.hh includes complex.h which defines the macro `I` which breaks HPX as I is used in templates.

File: test/unit/factorization/mc/test_compute_t_factor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rasolca) [reply test/unit/factorization/mc/test_compute_t_factor.cpp:380]: If nothing can change no check is needed.

File: include/dlaf/factorization/qr/t_factor_mc.h [hunk 5]
```diff
@@ -220,36 +226,40 @@ void QR_Tfactor<Backend::MC, Device::CPU, T>::call(const SizeType k, Matrix<cons
     // A possible solution to this would be to have multiple places where to store partial
     // results, and then locally reduce them just before the reduce over ranks
     // t = gemvColumnT(is_v0, k, v.read(v_i), taus, &t);
```
Comment (Sely85) [inline include/dlaf/factorization/qr/t_factor_mc.h:228]: Can this line be removed?

**Review by rasolca — APPROVED** [review]
